compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -1912,7 +1957,8 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         // (Only the last candidate can possibly have more match pairs.)
         debug_assert!({
             let mut all_except_last = candidates_to_expand.iter().rev().skip(1);
-            all_except_last.all(|candidate| candidate.match_pairs.is_empty())
+            all_except_last
+                .all(|candidate| matches!(candidate.state, CandidateState::Complete { .. }))
```

Comment:
It looks like this is overly strict, because we can also have `CandidateState::Expanded` with `remaining_match_pairs.is_empty()`.

That seems to be the cause of the assertion failure in debug-assertion builds.